experimental/zerotrie/src/builder/konst/store.rs

```diff
@@ -0,0 +1,294 @@
+
+    pub const fn from_manual_slice(full_slice: &'a [T], start: usize, limit: usize) -> Self {
+        ConstSlice {
+            full_slice,
+            start,
+            limit,
+        }
+    }
+
+    pub const fn len(&self) -> usize {
+        self.limit - self.start
+    }
+
+    pub const fn get_or_panic(&self, index: usize) -> &T {
+        &self.full_slice[index + self.start]
+    }
+
+    #[cfg(test)]
+    pub const fn first(&self) -> Option<&T> {
+        if self.len() == 0 {
+            None
+        } else {
+            Some(self.get_or_panic(0))
+        }
+    }
+
+    pub const fn last(&self) -> Option<&T> {
+        if self.len() == 0 {
+            None
+        } else {
+            Some(self.get_or_panic(self.len() - 1))
+        }
+    }
+
+    #[cfg(test)]
+    pub const fn get_subslice_or_panic(
+        &self,
+        new_start: usize,
+        new_limit: usize,
+    ) -> ConstSlice<'a, T> {
+        assert!(new_start <= new_limit);
+        assert!(new_limit <= self.len());
+        ConstSlice {
+            full_slice: self.full_slice,
+            start: self.start + new_start,
+            limit: self.start + new_limit,
+        }
+    }
+
+    #[cfg(any(test, feature = "alloc"))]
+    pub fn as_slice(&self) -> &'a [T] {
+        &self.full_slice[self.start..self.limit]
+    }
+}
+
+impl<'a, T> From<&'a [T]> for ConstSlice<'a, T> {
+    fn from(other: &'a [T]) -> Self {
+        Self::from_slice(other)
+    }
+}
+
+/// A const-friendly mutable data structure backed by an array.
+#[derive(Debug, Copy, Clone)]
+pub(crate) struct ConstArrayBuilder<const N: usize, T> {
+    full_array: [T; N],
+    start: usize,
+    limit: usize,
+}
+
+impl<const N: usize, T: Default> Default for ConstArrayBuilder<N, T> {
+    fn default() -> Self {
+        Self::new_empty([(); N].map(|_| Default::default()), 0)
+    }
+}
+
+impl<const N: usize, T> ConstArrayBuilder<N, T> {
+    pub const fn new_empty(full_array: [T; N], cursor: usize) -> Self {
+        assert!(cursor <= N);
+        Self {
+            full_array,
+            start: cursor,
+            limit: cursor,
+        }
+    }
+
+    pub const fn from_manual_slice(full_array: [T; N], start: usize, limit: usize) -> Self {
+        assert!(start <= limit);
+        assert!(limit <= N);
+        Self {
+            full_array,
+            start,
+            limit,
+        }
+    }
+
+    pub const fn len(&self) -> usize {
+        self.limit - self.start
+    }
+
+    #[allow(dead_code)]
+    pub const fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub const fn as_const_slice(&self) -> ConstSlice<T> {
+        ConstSlice::from_manual_slice(&self.full_array, self.start, self.limit)
+    }
+
+    #[cfg(feature = "alloc")]
+    pub fn as_slice(&self) -> &[T] {
+        &self.full_array[self.start..self.limit]
+    }
+}
+
+impl<const N: usize> ConstArrayBuilder<N, u8> {
+    pub const fn const_bitor_assign(mut self, index: usize, other: u8) -> Self {
```

Comment:
suggestion: I'd call the parameter `bits`

Reply:
I like it; done